content/python/upgrade_virtual_guest_dedicated.py.md

```diff
@@ -0,0 +1,421 @@
+description: "A few examples on interacting with upgrade virtual guest."
+date: "2019-03-27"
+classes:
+    - "SoftLayer_Virtual_Guest"
+    - "SoftLayer_Product_Order"
+    - "SoftLayer_Container_Product_Order_Virtual_Guest_Upgrade"
+tags:
+    - "virtualserver"
+---
+
+Upgrade a VSI(cpu, ram)
+```
+import json
+
+import SoftLayer
+
+client = SoftLayer.create_client_from_env()
+orderService = client['SoftLayer_Product_Order']
+
+# The id of the virtual guest you wish upgrade
+virtualGuestId = 11111
+
+# Package keyName.
+package_keyname = 'CLOUD_SERVER'
+
+object_mask = 'mask[prices]'
+
+# items KeyNames to be upgrade
+items = ['GUEST_PRIVATE_CORES_8', 'RAM_12_GB']
+
+response_package = client["SoftLayer_Product_Package"].getAllObjects()
+
+package_id = None
+for val in response_package:
+    if val.get('keyName') == package_keyname:
+        package_id = val.get('id')
+
+get_item_id_price = client["SoftLayer_Product_Package"].getItems(mask=object_mask, id=package_id)
+
+prices = []
+for keyname_item in items:
+    for get_item in get_item_id_price:
+        if keyname_item == get_item.get('keyName'):
+            prices.append(get_item['prices'][0]['id'])
+
+# Build a skeleton SoftLayer_Container_Product_Order_Property objects
+properties = [
+    {
+        "name": "MAINTENANCE_WINDOW",
+        "value": "2018-08-20T06:04:10Z"},
+    {
+        "name": "NOTE_GENERAL",
+        "value": "Upgrade instance configuration."
+
+    }
+]
+
+# Build a skeleton SoftLayer_Virtual_Guest object to model the id
+virtualGuests = [
+    {
+        "id": virtualGuestId
+
+    }
+]
+
+"""
+Build a skeleton SoftLayer_Container_Product_Order_Virtual_Guest_Upgrade object
+containing the upgrade you wish to place.
+"""
+orderData = {
+    "complexType": "SoftLayer_Container_Product_Order_Virtual_Guest_Upgrade",
+    'packageId': package_id,
+    'prices': [{'id': price_id} for price_id in prices],
+    'properties': properties,
+    'virtualGuests': virtualGuests
+}
+
+try:
+    # Upgrading the virtual guest
+    response = orderService.verifyOrder(orderData)
+    print(json.dumps(response, sort_keys=False, indent=4, separators=(',', ': ')))
+
+except SoftLayer.SoftLayerAPIError as e:
+    """
+        If there was an error returned from the SoftLayer API then bomb out with the
+        error message.
+        """
+    print("Unable to upgrade VSI \nfaultCode= %s, \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+
+
+```
+
+Upgrade a VSI adding two new disks.
+```
+import json
+
+import SoftLayer
+
+client = SoftLayer.create_client_from_env()
+
+mask_prices = 'mask[prices[categories]]'
+
+# The id of the virtual guest you wish upgrade
+virtualGuestId = 11111
+
+# Package keyName.
+package_keyname = 'CLOUD_SERVER'
+
+# items KeyNames to be upgrade
+items_list = ['GUEST_DISK_10_GB_SAN',  # "description": "10 GB (SAN)", "categoryCode": "guest_disk1"
+                  'GUEST_DISK_20_GB_SAN'  # "description": "20 GB (SAN)", "categoryCode": "guest_disk2"
```

Comment:
I'm sure they will be added in the same order but what happens if user wants this GUEST_DISK_20_GB_SAN at 4th position? 

I think you can improve this example or provide an explanation about how to specify the position of the disk

Reply:
To upgrade a disk on a vs the price id for this item  can be set as "categoryCode": "guest_disk1", "categoryCode": "guest_disk2", etc. This have to be set in the json body "prices" as in the script, you can see it below.

```
#  keyName = GUEST_DISK_20_GB_SAN
'prices': [{'id': prices[0], 
                'categories': [
                    {
                        'categoryCode': 'guest_disk4',
                        'complexType': 'SoftLayer_Product_Item_Category'
                    }
                ],
                },

```

Below you can see a item price with their "categories" options:

```
{
        "id": 152377,
        "item": {
            "description": "10 GB (SAN)",
            "id": 1213,
            "keyName": "GUEST_DISK_10_GB_SAN",
            "categories": [
                {
                    "categoryCode": "guest_disk2",
                    "id": 92,
                    "name": "Third Disk",
                },
                {
                    "categoryCode": "guest_disk3",
                    "id": 93,
                    "name": "Fourth Disk",
                },
                {
                    "categoryCode": "guest_disk4",
                    "id": 116,
                    "name": "Fifth Disk",
                },
```